Fixes #10487: Add custom logging support.

ehelms opened the pull request:
This change moves away from the Rails logger to the 'logging' gem
to support more configurability of logging. Further, this allows for
the creation of custom loggers that plugins can then register through
settings.

For example, the plugin katello could define within its settings file
(settings.plugins.d/katello.yaml) the following:

```
:katello:
  :loggers:
    :pulp_rest:
      :enabled: false
    :cp_rest:
      :enabled: true
```

Which would define 2 loggers, pulp_rest and cp_rest where the latter
is turned on be default, but the former can be enabled such that
the log file isn't continuously filled with messages from that particular
layer.

ehelms committed:
Fixes #10487: Add custom logging support.

This change moves away from the Rails logger to the 'logging' gem
to support more configurability of logging. Further, this allows for
the creation of custom loggers that plugins can then register through
settings.

For example, the plugin katello could define within its settings file
(settings.plugins.d/katello.yaml) the following:

```
:katello:
  :loggers:
    :pulp_rest:
      :enabled: false
    :cp_rest:
      :enabled: true
```

Which would define 2 loggers, pulp_rest and cp_rest where the latter
is turned on be default, but the former can be enabled such that
the log file isn't continuously filled with messages from that particular
layer.

ohadlevy commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,169 @@
+      rescue Errno::EACCES
+        # just notify insufficient privileges
+        # rubocop:disable Output
+        puts "Insufficient privileges for #{@log_directory}"
```
warn ?

ares replied:
I think logger is not ready at this moment so warn can't be used

ohadlevy replied:
I mean Kernel.warn

see http://apidock.com/ruby/Kernel/warn

ares replied:
ah right, makes sense

ohadlevy commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,169 @@
+require 'logging'
+
+module Foreman
+  class LoggingImpl
```
why Impl vs simple Foreman::Logger ?

ehelms replied:
That is just a copy over from the original implementation in Katello. In the end, we could go either way as it's largely inconsequential given https://github.com/theforeman/foreman/pull/2376/files#diff-6bb962bdbc2f8a8c1ea2df38f09f8890R168 such that the object present in the environment is `Foreman::Logging`.

ohadlevy commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,169 @@
+      fail 'logging can be configured only once' if @configured
+      @configured = true
+      @log_directory = options.fetch(:log_directory, './log')
+      @config = YAML.load_file('config/logging.yaml')
```
need to handle if the file doesnt exists etc

ohadlevy replied:
somehow this comment got lost in github, please review

ohadlevy commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,169 @@
+    def build_syslog_appender(name, options)
+      ::Logging.appenders.syslog(
+        name,
+        options.reverse_merge(:ident    => "#{options[:prefix]}katello",
```
leftover ?

ohadlevy commented on config/logging.yaml:
```diff
@@ -0,0 +1,46 @@
+    # you log files will be stored here
+    # note when you log outside of standard rails path you won't see log output in WEBrick
+    # STDOUT (usually not wanted in development), to use Rails default path, don't set path
+    # path: /var/log/katello
```
s/katello/foreman ?

ohadlevy commented:
missing tests?

ohadlevy commented:
[test] seems like jenkins oom failures

domcleal commented:
> seems like jenkins oom failures

That's because it's generating a 150MB log file from each test run.  Please don't re-run it until this is fixed.

The tests are outputting all debug level logs to stdout or stderr, which is generating a huge amount of data, can that be disabled?

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -93,6 +93,17 @@ def installed?(id)
+    def add_loggers
+      return unless (defined?(Foreman::Logging) && SETTINGS[@id])
+
+      if (loggers = SETTINGS[@id][:loggers])
```
If a plugin tries to use a logger X that doesn't exist in SETTINGS, what happens?

ehelms replied:
AFAIK this would be the same condition as if a logger is disabled (i.e. nothing would get output to the log file).

ehelms replied:
An example of this we use today is for Pulp rest call logging. We hand the Runcible library that makes our calls down to Pulp a logger and then can disable/enable it (or if undefined, nothing would happen) - https://github.com/Katello/katello/blob/master/app/lib/katello/util/thread_session.rb#L67

One question I am not sure on is whether plugins should be able to define per environment settings for their loggers. Per environment settings was something I missed in the first pass of this implementation (and resulted in the massive log files and console printing) and will be accounted for in the next push.

domcleal replied:
Seems like we need a mechanism for default configuration which doesn't rely on SETTINGS, or a way to prepopulate SETTINGS with that default configuration, else plugins aren't going to be configured well by default - especially on upgrade if config files aren't updated.

It's also a different configuration mechanism to Foreman core which appears to now use the logging.yaml file, which could be confusing for users.

ehelms replied:
For Foreman core, I considered putting the configuration into the settings.yaml file, but due to the size of the default configuration I thought a separate yaml file more consumable (and not crowd the settings.yaml file). If folks would prefer I can rework this to have `:logging:` node in settings.yaml with what is in logging.yaml now.

When you say default configuration - are you referring to default configurations for plugins or for logging in general?

domcleal replied:
I'm referring to default configuration for logging in plugins, as that's what this seems to depend upon.  It seems that if the settings file is older or not present then we won't have any good defaults for plugin logging as their loggers won't have been configured.

I do agree about keeping Foreman core's logging in a separate config file, that seems like a good idea, but it does mean one rule for core and one for plugins, which isn't good for consistency.

Perhaps break the configuration into two, with the configuration of stuff like paths and rotation in a logging.yaml as that applies to both core+plugins, the per-logger levels in settings files and ensure that both core+plugins can set sensible defaults for their loggers in code without requiring those settings files.

(Sorry if this is a bit rambly.)

ehelms replied:
Not rambly at all, its good feedback to consider.

For the defaults, I am not able to think what defaults a plugin would need to worry about out of the gate. If the plugin is using things like 'Rails.logger' then they will continue to log to the same place and in the same way as before this patch. The log messages will show up as 'app' in the log files. 

If we want to, for the purpose of making logs easier to read, start to enforce per-plugin logs then we would need some default log creation if not specified within the settings file for the plugin. This could be based on the Rails engine name for consistency across the board. Note that this would only apply to specifically inserted log messages (Rails will continue to log requests and such under the 'app' logger). For example,

```
Katello.logger = ::Logging.logger['katello']
Katello.logger.debug "Debugging message for this katello action"
```

Which would then appear within the logs as:

```
2015-05-13 13:28:22 [katello] [D] Debugging message for this Katello action
```

I think splitting the config and the definition of the loggers themselves makes sense across settings and logging. I would expect then, that the 'environmental' configurations would also live under settings (https://github.com/theforeman/foreman/pull/2376/files#diff-d71db24719150ef52a8afd6ef934f9b4R54)?

domcleal replied:
> For the defaults, I am not able to think what defaults a plugin would need to worry about out of the gate. If the plugin is using things like 'Rails.logger' then they will continue to log to the same place and in the same way as before this patch.

It was the custom loggers I was thinking about, so if a plugin decided to use a custom logger to make some of its log messages more categorised, and assuming they should be logged by default, we should ensure the logger is set up even without an explicit config file.

> If we want to, for the purpose of making logs easier to read, start to enforce per-plugin logs then we would need some default log creation if not specified within the settings file for the plugin. This could be based on the Rails engine name for consistency across the board.

I think that'd be a great idea.

> I think splitting the config and the definition of the loggers themselves makes sense across settings and logging. I would expect then, that the 'environmental' configurations would also live under settings (https://github.com/theforeman/foreman/pull/2376/files#diff-d71db24719150ef52a8afd6ef934f9b4R54)?

Yeah, I guess if the rotation settings aren't there then we minimise what's in the general settings file.  My aim is to keep the commonly changed things (log levels) away from rarely changed things (formatting/rotation etc) so misconfigurations (or rather reusing old config files) are less likely to cause bad or missing logging.

lzap commented on config/logging.yaml:
```diff
@@ -0,0 +1,46 @@
+    #  [%X{string}] variable set using ::Logging.mdc['string'] =
+    :pattern: "%d [%c] [%.1l] %m\n"
+    # log filename for this environment, it will be placed into #{Rails.root}/log directory
+    :filename: "production.log"
```
I think this is the great chance of getting rid of log files completely and sending everything through syslog/journald. What folks think about defaulting to syslog and providing syslog configuration to file all foreman data into the correct filename?

lzap replied:
Or if you feel that syslog by default is too strong, than let's introduce some setting that will easily configure it with syslog. Just one flag `syslog` yes/no is easy for us to configure in packages.

ehelms replied:
Testing configuring it to syslog worked (and pointed out that it wasn't logging it as the 'foreman' user properly). Out of the box it looked like:

```
May 13 16:00:25 katello-develnew foreman[21787]: INFO  app : Processing by UsersController#login as HTML
May 13 16:00:25 katello-develnew foreman[21787]: INFO  app :   Rendered users/login.html.erb within layouts/login (76.8ms)
May 13 16:00:25 katello-develnew foreman[21787]: INFO  app :   Rendered layouts/base.html.erb (317.2ms)
```

Discussion worthy for production; however, I am not a syslog master.

lzap replied:
Cool. My main concern was if you can simply provide some comments in the configuration file to get that setup. How do I achieve that setup?

lzap commented on config/application.rb:
```diff
@@ -136,6 +137,12 @@ class Application < Rails::Application
+    config.colorize_logging = Foreman::Logging.config[:colorize]
+
+    config.logger = ::Logging.logger['app']
+    config.active_record.logger = ::Logging.logger['sql']
```
I really wish this was logger destination per active record model as I hate Taxonomies and Permissions SQL logs most of the time... ;-)

ehelms replied:
There is a rails-logging gem that does this sort of (https://github.com/TwP/logging-rails#usage) on a per file basis but I thought it might be overkill.

lzap commented on config/environment.rb:
```diff
@@ -1,12 +1,5 @@
-  def format_message(severity, timestamp, progname, msg)
-    "#{timestamp.strftime('%Y-%m-%d %H:%M:%S')} [#{severity[0]}] #{msg}\n"
-  end
-end
```
Can we keep this format for the new logger? Or at least make sure the output is not that verbose (short year, severity short string) as we want to save as many terminal columns as possible for the actual data.

ehelms replied:
If you look at https://github.com/theforeman/foreman/pull/2376/files#diff-d71db24719150ef52a8afd6ef934f9b4R36 you'll see that I kept this same format but added in the name of the logger just before the severity. e.g.

```
2015-05-13 13:28:22 [app] [I] Connecting to database specified by database.yml
The Apipie cache is turned off. Enable it and run apipie:cache rake task to speed up API calls.
2015-05-13 13:28:33 [sql] [D]   Setting::General Load (0.9ms)  SELECT "settings".* FROM "settings" WHERE "settings"."category" IN ('Setting::General') AND "settings"."name" = 'fix_db_cache' ORDER BY name LIMIT 1
```

lzap commented on config/logging.yaml:
```diff
@@ -0,0 +1,46 @@
+# was this message logged
+:log_trace: false
+#do not log certain paths (example  /notices/get_new)
+:ignored_paths: []
```
What do you mean by path? Are these the logging destinations like `pulp_rest` you mentioned in the description? Then perhaps we should either adopt the slash convention, or change this comment including some more valid examples.

ehelms replied:
No, I believe this is so that certain paths can be ignored such as `/assets/` that crowd the logs with unneeded information.

lzap commented on config/logging.yaml:
```diff
@@ -0,0 +1,46 @@
+  # which is used by all other loggers
+  :root:
+    # only messages with severity >= this level from all loggers will be logged
+    :level: debug
```
Can you please list all the severity levels in the comment? One never remembers if these are called "warn" or "warning" etc.

Also I'd really appreciate we have default level of `debug` when in development mode and `warning` when in production mode. Therefore this line would be commented out by default.

lzap replied:
All my concerns are resolved, except this one. I'd appreciate list of values which are allowed here.

lzap commented on config/logging.yaml:
```diff
@@ -0,0 +1,46 @@
+  :app:
+    :enabled: true
+  :sql:
+    :enabled: true
```
I think I don't understand the comment and the last block just by reading this config file. Can you explain this in more detail please?

ares replied:
If you talk about the `path:` comment, there's one typo (s/you/your/) and the meaning is that if you run your app in webrick and change the path, you won't see logs in webrick's STDOUT, because webrick just `tail -f ./logs/$env.log`. So in development it's not a good idea to change the log path. If you don't want to override the default path that Rails sets (./logs) just don't uncomment path option.

ehelms replied:
Looks like I also missed a comment from the original implementation about the loggers block. I will add that back and update the comment.

lzap commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,169 @@
+      logger.level    = logger_config[:level] if logger_config.key?(:level)
+      logger.additive = logger_config[:enabled] if logger_config.key?(:enabled)
+      logger.trace    = @config[:log_trace]
+      enable_tailing(logger, logger_config) if logger_config.key?(:tail_command)
```
This option is undocumented (no example and comment in the default config).

Speaking about tailing, would it be too insane to include a simple API to tail logs from the Rails application itself, so we can create a tab in the About box to display last N lines of logs? :-)

ehelms replied:
This appears to be something custom that was added in the original implementation. Looking at the method, I think it allowed for including tailing some other log (e.g. messages) that would then get mixed into the production log. I am going to remove this unless someone speaks up with a valid use case.

lzap commented:
Big improvement, I like it a lot!

ehelms commented:
One design question I have been debating is how we handle user configuration in the event they want to escalate a log level. If `logging.yaml` is not deployed by the puppet module then users do not run a risk of losing their settings. However, for plugins that deploy and configure their settings files with the installer, the changed settings being over written is a possibility.

ohadlevy commented on config/logging.yaml:
```diff
@@ -0,0 +1,83 @@
+    # log filename for this environment, it will be placed into #{Rails.root}/log directory
+    # note when you log outside of standard rails path you won't see log output in WEBrick
+    # STDOUT (usually not wanted in development), to use Rails default path, don't set path
+    # path: /var/log/katello
```
update path or keep the katello one?
I would prefer to see a unified log base directory?

ehelms replied:
Good catch, I fixed it, then copy pasted a block again. Will update.

domcleal commented on config/logging.yaml:
```diff
@@ -0,0 +1,83 @@
+    # pattern or rolling options
+    :type: file
+    # 'daily', 'weekly', 'monthly' or number as seconds
+    :age: 'weekly'
```
Currently we use logrotate for aging of logs, can/should we continue doing that?  I suspect it's more flexible, for instance, it's configured to compress logs.

ehelms replied:
It does not appear that the logging gem currently supports compression of log files when rotating them. I can remove this configuration aspect and the rolling_file appender which will then allow this to be handled by logrotate.

domcleal replied:
Yeah, that's my preference I think, as long as it handles reopening of the log file after rotation without the need for signals (or we can add a signal).

domcleal commented on config/logging.yaml:
```diff
@@ -0,0 +1,83 @@
+    :filename: "production.log"
+
+  # you can configure each logger individually, you can turn it off completely by using key
+  # 'enabled' or you can adjust it's logging 'level'. If you specify level higher than root
```
`its`

domcleal commented on config/logging.yaml:
```diff
@@ -0,0 +1,83 @@
+      :level: fatal
+
+:development:
+  :ignored_paths: ['/assets/']
```
Does this obsolete quiet_assets?

ehelms replied:
Looks like this was actually a custom solution for allowing ignoring any paths through middleware -- https://github.com/Katello/katello/blob/master/lib/katello/middleware/silenced_logger.rb . If support for this generic ignoring is something we want I can port this middleware over, otherwise I can remove this aspect and keep quiet_assets.

domcleal replied:
Perhaps keep quiet_assets for now until there's more of a need for it, then we can swap.

domcleal commented on config/application.rb:
```diff
@@ -136,6 +137,12 @@ class Application < Rails::Application
     config.middleware.use "Apipie::Middleware::ChecksumInHeaders"
 
+    Foreman::Logging.configure(:log_directory => "#{Rails.root}/log", :environment => Rails.env)
+    config.colorize_logging = Foreman::Logging.config[:colorize]
```
Note that we also set colorize_logging in all three config/environments/\* files based on SETTINGS.  Is that obsolete?

ehelms replied:
Yea, I think this would obsolete it as we would want to control it from a central configuration for logging. I will update these instances, but I assume I will also need to update the puppet module once we settle on the overall configuration setup.

domcleal replied:
Our module doesn't cover logging today, so we can do that at a later stage.  It should work correctly by default without any use of the installer.

domcleal commented on config/logging.yaml:
```diff
@@ -0,0 +1,83 @@
+  :root:
+    # only messages with severity >= this level from all loggers will be logged
+    # valid values:  debug, info, warn, error, fatal, unknown
+    :level: warn
```
normally we have info level logs in production, which show requests at least, is that still the case?

ehelms replied:
I will update this to info.

ehelms commented:
These test failures are strange in that I do not get them when I run the tests locally.

ehelms commented:
Still trying to figure out why these Subnet tests fail here but not locally for myself and will be adding unit testing for the logging class itself. However, I wanted to push out the redesign of configuration and plugin log handling to get feedback.

I adopted a style similar to database.yaml for logging.yaml by splitting configuration across environments. These values equate to how the root logger, and thus how child loggers are configured unless they specify overrides. Then, in settings.yaml the actual loggers (app and sql at this point) are declared.

For plugin logging, you can see (https://github.com/theforeman/foreman/pull/2376/files#diff-fe65def4d8ec1ce9d44c8b067c24cbdbR4) for an explanation of the design around how they are configured, how the default is created and subsequent loggers are namespaced.

One question I had was whether this made sense to do  (https://github.com/theforeman/foreman/pull/2376/files#diff-6bb962bdbc2f8a8c1ea2df38f09f8890R40) which is basically to throw a fail message if a plugin attempts to use an undefined logger. My thinking is that loggers should be configured and using the system we have in place for namespacing them so that the logs aren't able to be crowded with "unknown" loggers or loggers that break namespace boundaries that developers have put into their code.

ehelms commented:
Added some tests for the Foreman::Logging class and added a check in application.rb to handle upgrades where the logging setup is not yet present.

Worth noting that the broken Subnet tests were due to this situation:

```
logger.debug("message") and return
```

Which doesn't work with the custom logging because the log methods return either true or false based on whether they generated a log message or not. This can happen if the logger is disabled, or the log level of the logger is set below that of the root logger.

domcleal commented on Gemfile:
```diff
@@ -27,6 +27,7 @@ gem 'fast_gettext', '~> 0.8'
 gem 'gettext_i18n_rails', '~> 1.0'
 gem 'i18n', '~> 0.6.4'
 gem 'turbolinks', '~> 2.5'
+gem 'logging', '>= 1.8.0', '< 2.0.0'
```
Is there anything in 2.0.0 that's a problem?  A brief scan of the changelog looks okay.  I ask as it's out, so I don't want to start on the back foot, and we're already using it in our bundles due to another gem using `~> 2.0`.

domcleal commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,159 @@
+      logger          = ::Logging.logger[logger_name]
+      logger.level    = logger_config[:level] if logger_config.key?(:level)
+      logger.additive = logger_config[:enabled] if logger_config.key?(:enabled)
+      logger.trace    = @config[:log_trace]
```
Answering my own question about 2.0.0, this causes a small problem here and on the root logger: https://github.com/TwP/logging/commit/5c06041

If we could support both, that'd be appreciated, until we can get rid of our older logging RPM.

domcleal commented on config/settings.yaml.example:
```diff
@@ -23,5 +23,14 @@
+    :enabled: true
+  :sql:
+    :enabled: true
+    :level: fatal
```
any need for this?  I'm trying to think if there's any AR logging we're trying to hide between info and fatal level, and can't think of anything - it's mostly all debug.

domcleal commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,159 @@
+          :error => :red,
+          :fatal => [:white, :on_red]
+        },
+        :date   => :blue,
```
At the risk of starting a bikeshed discussion, this is rather unreadable for me on a black terminal.

domcleal commented on config/application.rb:
```diff
@@ -136,6 +137,16 @@ class Application < Rails::Application
+
+    # Check that the loggers setting exist to configure the app and sql loggers
+    # This is required for upgrades
+    if SETTINGS[:loggers]
```
Could we do this automatically if the config's missing?

ehelms replied:
To clarify, create the two loggers app and sql if the config is missing?
On May 21, 2015 7:02 AM, "Dominic Cleal" notifications@github.com wrote:

> In config/application.rb
> https://github.com/theforeman/foreman/pull/2376#discussion_r30791852:
> 
> > @@ -136,6 +137,16 @@ class Application < Rails::Application
> >      # Add apidoc hash in headers for smarter caching
> >      config.middleware.use "Apipie::Middleware::ChecksumInHeaders"
> > -    Foreman::Logging.configure(:log_directory => "#{Rails.root}/log", :environment => Rails.env)
> > -    Foreman::Logging.add_loggers(SETTINGS[:loggers])
> >   +
> > -    # Check that the loggers setting exist to configure the app and sql loggers
> > -    # This is required for upgrades
> > -    if SETTINGS[:loggers]
> 
> Could we do this automatically if the config's missing?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/2376/files#r30791852.

domcleal replied:
Yep.

domcleal commented on config/settings.yaml.example:
```diff
@@ -23,5 +23,14 @@
+# Loggers can be configured individually, they can be turned off completely by using key
+# 'enabled' or you can adjust its logging 'level'. If you specify a level higher than root
+# logger has, messages below this level are ignored. If you specify lower level than root
+# they are still ignored but you can add appender in Foreman::Logging just for this logger
```
This doesn't seem like very useful advice for users... is there any way to make this work by default, or should we just recommend that they keep the root logger level below any other debug they want?

domcleal commented on lib/foreman/logging.rb:
```diff
@@ -0,0 +1,159 @@
+                                :layout   => build_layout(@config[:pattern], @config[:colorize]))
+        )
+      rescue ArgumentError
+        # if appender cannot open output file we ignore it, STDOUT fallback will be used
```
Should this be mentioned on stderr etc?

domcleal commented on test/unit/plugin_logging_test.rb:
```diff
@@ -0,0 +1,46 @@
+  def assert_logger_exists(name)
+    assert ::Logging::Repository.instance.has_logger?(name)
+  rescue Minitest::Assertion
+    raise Minitest::Assertion, "Failed assertion, logger #{name} does not exist"
```
I don't think I've seen this re-raising pattern with assertions before - can't you just add the message as the second parameter to `assert` above?

domcleal commented:
I might be misremembering, but I thought there was something mentioned earlier about being able to have logs from a plugin going into the namespaced plugin logger by default.  I don't see any change to the engine's default app logger, if that's even possible?

The PR also needs rebasing.

Could you add a test to test/unit/plugin_test.rb to check the configure_logging is happening when a plugin is registered?

> One question I had was whether this made sense to do (https://github.com/theforeman/foreman/pull/2376/files#diff-6bb962bdbc2f8a8c1ea2df38f09f8890R40) which is basically to throw a fail message if a plugin attempts to use an undefined logger

This seems OK for now, I can't think of a reason not to do it as long as we keep configuring any loggers we need in Foreman core by default and in the example settings file, and plugins use the new API.

ehelms commented:
> I might be misremembering, but I thought there was something mentioned
> earlier about being able to have logs from a plugin going into the
> namespaced plugin logger by default. I don't see any change to the engine's
> default app logger, if that's even possible?

If a plugin is currently only using 'Rails.logger' or 'logger' depending on
where they are in the code base (e.g. an ActiveRecord model) the messages
would go to the app/sql loggers. I might be able to achieve this by
overriding the various logging methods (e.g. debug, warn, etc.) to try to
do some inspection of the call stack to determine the module that called
the function but that sounds quite magical. Or if someone knows a Ruby way
when a function is called to easily determine the original callers class
that I could then use to check if a known plugin called it. Outside of
that, I believe plugins would have to convert themselves over to using
syntax like:

```
Foreman::Logging.logger('katello').debug "My debug message"
```

domcleal commented:
> If a plugin is currently only using 'Rails.logger' or 'logger' depending on where they are in the code base (e.g. an ActiveRecord model) the messages would go to the app/sql loggers.

Yeah, it was the latter use (within a model/controller) I had expected that we could change somehow.  Perhaps we could fix it in controllers by overriding or setting the `logger` attr, as they're all likely to extend our own base controllers, but models would be harder.  Oh well, something to think about for another time I suppose.

ehelms commented:
> If a plugin is currently only using 'Rails.logger' or 'logger' depending
> on where they are in the code base (e.g. an ActiveRecord model) the
> messages would go to the app/sql loggers.
> 
> Yeah, it was the latter use (within a model/controller) I had expected
> that we could change somehow. Perhaps we could fix it in controllers by
> overriding or setting the logger attr, as they're all likely to extend
> our own base controllers, but models would be harder. Oh well, something to
> think about for another time I suppose.

One potential way we could do this is to create a mixin that is included on
load in every ActiveRecord object (the same _could_ be done for controller
to as well). For example, during initialization we do something like:

```
ActiveSupport.on_load(:active_record) { self.__send__(:include,
Foreman::Logging::Mixing) }
```

domcleal commented:
Yeah, and then use the model namespace, location of the file or something to identify the plugin.  I'm happy to treat that as a separate ticket, as it'd need some thought - we can build on top of what's here.

ehelms commented:
2 final design questions:

1) Do we see any reason to support logger specific log files? For example, the 'sql' logger could declare a log file for use in its configuration and the logging setup would configure said logger to log only to that file or both to the root log and the declared log file.

2) With this design, if a user wants to bump the logging level for all loggers from say info to debug, they would need to edit logging.yaml and change it there, then restart the application. If they wanted to bump a single logger, say 'sql' to debug while the rest stayed at info, they would need to edit settings.yaml and change that loggers config level. Is this workflow ok?

ehelms commented:
Actually, I should add a third I suppose:

3) Do we want to support syslog in production? or the ability to do both syslog and file in production?

domcleal commented on Gemfile:
```diff
@@ -27,6 +27,7 @@ gem 'fast_gettext', '~> 0.8'
 gem 'gettext_i18n_rails', '~> 1.0'
 gem 'i18n', '~> 0.6.4'
 gem 'turbolinks', '~> 2.5'
+gem 'logging', '>= 1.8.0'
```
and `< 3.0.0` perhaps for safety, they seem to follow a semver-like style

domcleal commented:
> 1) Do we see any reason to support logger specific log files?

I don't really see a use at the moment.

> 2) With this design, if a user wants to bump the logging level for all loggers from say info to debug, they would need to edit logging.yaml and change it there, then restart the application. If they wanted to bump a single logger, say 'sql' to debug while the rest stayed at info, they would need to edit settings.yaml and change that loggers config level. Is this workflow ok?

Yeah, that's ended up a little oddly.  What do you think, maybe move global level to settings.yaml too?

> 3) Do we want to support syslog in production? or the ability to do both syslog and file in production?

Yeah, we should, but probably syslog-only rather than both.

ehelms commented:
Updated with:

1) Production environment set to 'syslog' by default
2) Logging settings can now be set for the current environment through settings.yaml (see https://github.com/theforeman/foreman/pull/2376/files#diff-c08285198f9b2708ddd77e0faef7baa8R26)
3) Tried 'green' for the date in console mode for contrast, if that is still ugly I will just update to white.

domcleal commented:
> 1) Production environment set to 'syslog' by default

Sorry, I guess I misunderstood earlier, I thought this was just to enable it as an option rather than to change the default.

I'm not sure with the amount we log and because it's multi-line that we should be doing this - we might be best keeping the file as default, then later on integrating to journald as an option.  Sorry for the confusion.

domcleal commented:
> 3) Tried 'green' for the date in console mode for contrast, if that is still ugly I will just update to white.

Works for me, thanks.

ehelms commented:
Updated to move back to file as the default for production.

domcleal commented on app/services/foreman/plugin/logging.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      def loggers
+        config = @config[:loggers] || {}
```
It looks like plugins might have the same issue you fixed for core, where the default loggers need to be defined by a settings file to work.  Can you add an API for plugins to be able to define their default loggers so they don't hit errors when trying to use something not defined by a config file?

ehelms commented:
I updated the way plugins are handled to allow defining loggers in the plugin definition or settings. Or, if the loggers are defined in the plugin definition, whatever is in the settings file will override. API for loggers would look like:

```
Foreman::Plugin.register :test_plugin do

  logger :test, :enabled => true
  logger :rest_api, :enabled => true

end
```

Which, in the same way, would result in two loggers:  `test_plugin/test` and `test_plugin/rest_api`

domcleal commented:
[test]

domcleal commented:
Looks good, thanks.

There seems to be a problem with the Katello tests, probably the pinning of logging to < 2.0.0 causing bundler some problems with our GCE bundler group.  I'd suggest loosening it to < 3.0.0 if possible to fix it.

Edit: I see you have a PR open, so shall I just merge this?

ehelms commented:
I updated my PR to Katello to make use of the new plugin API and fix some tests. My run of the two PRs together passed and it is in review. So I'd say when you want to merge this you can. I'll follow this up with an update to the plugin wiki page describing custom loggers.

domcleal commented:
Merged as 3088e64, thanks @ehelms.

The wiki update would be much appreciated, cheers.
